### ISSUE TITLE
[DT Master 4.1.0] color balance rgb perceptual briliance grading highlight slider creates artifacts

### ISSUE THREAD

**Issue by s7habo** [darktable-org/darktable#12442]
**Describe the bug/issue**

If the highlights of the photo are close to the limit of dynamic range, the highlight slider of the perceptual briliance grading in the color balance rgb module produces artifacts when a certain threshold is exceeded. 

This only happens in the default _darktable USC (2022)_ color space. In JzAzBz (2021) color space the slider behaves as expected. 

**To Reproduce**

1.  increase the exposure of the photo with exposure module until the highlights are at the limit of the dynamic range.
2. in the color balance rgb module, increase the value of perceptual brilliance in highlights using the highlights slider. 
3. After a certain value, the artifacts - wide white spots - appear (see screenshot/video below).

**Screenshots**

![Artifact](https://user-images.githubusercontent.com/38165484/189203438-3bfec8e8-b120-48c1-b5e0-4d2a1d719115.jpg)

**Screencast**

Short demo:

https://user-images.githubusercontent.com/38165484/189202459-4b598eb2-263c-421e-ba1c-e04402929b3e.mp4

**Platform**
_Please fill as much information as possible in the list given below. Please state "unknown" where you do not know the answer and remove any sections that are not applicable _

* darktable version : 4.1.0~git272.5a1a1845-1
* OS                : Linux
* Linux - Distro    :  Ubuntu Studio 22.04
* Memory            : 31,3 GiB
* Graphics card     : NVIDIA GeForce GTX 1070/PCIe/SSE2
* Graphics driver   : 470.141.03
* OpenCL installed  : yes
* OpenCL activated  : yes
* Desktop           : KDE-Plasma-Version: 5.24.6

**Additional context**

I can't say for sure, but I suspect that this problem may be related to the 

https://github.com/darktable-org/darktable/pull/12233 

since this behavior appeared after that commit.

**Comment by gi-man** [comment]
I can reproduce too with and without OpenCL. 4.1.0~git296.1ad26dc5 Fedora 36, X11

**Comment by todd-prior** [comment]
I think AP generates this in the last 10 min of his most recent video and he comments that it is due to gamut mapping....

**Comment by Mark-64** [comment]
I have this too and I believe that it is generated in filmic, maybe by reconstruction kickin-in at a certain point when increasing exposure or brialliance.
Reconstruction is never really disabled, the threshold is set high instead, but it can alway be activated if exposure is set high enough

**Comment by pehar1** [comment]
@Mark-64 , this has already been discussed here : https://discuss.pixls.us/t/something-wrong-with-darktable-4-1-0-git272-5a1a1845-1-clipped-highlights/32540

**Comment by pass712** [comment]
Can reproduce under Windows.

dt: 4.1.0+283~g5cca9d64b
OS: Win10
Mem: 16 Gb
Graphics card: NVidia GEFORCE GTX 1060 6GB
OpenCL installed : yes
OpenCL activated : yes

**Comment by jenshannoschwalm** [comment]
As @s7habo hinted the problem is somewhat known and discussed in the mentioned pr.
Pinging @flannelhead here...

**Comment by flannelhead** [comment]
The fix that was done in https://github.com/darktable-org/darktable/pull/12233 was about avoiding divisions by zero.

Here's the result with current master:
![Screenshot from 2022-09-09 19-36-23](https://user-images.githubusercontent.com/5001906/189401560-6a54aa04-7bcd-4225-b6f7-dcc56908fc54.png)
With that PR reverted:
![Screenshot from 2022-09-09 19-49-55](https://user-images.githubusercontent.com/5001906/189401628-94efe1d7-91d2-4101-b8d0-64e7773c712f.png)
Notice that in the latter image there are specular highlights with black pixels. That's what the fix meant to address. However, it seems here that neither result is really acceptable.

As others have concluded here, the large blotches are caused by filmic's reconstruction which seems to spread those specular highlights. color balance rgb's brilliance adjustment in dt UCS mode seems to push specular highlights quite far, so filmic's reconstruction picks up these very bright highlights and spreads them all around.

Contrary to what some stated about filmic's reconstruction, at least in the case of this image it can be practically disabled. Turn the threshold all the way up to 6 EV and transition slider all the way down to 0.25. At least with those settings I could no longer reproduce this sort of artefacts.

It might be worth investigating why color balance rgb brilliance adjustment causes those insanely high values for specular highlights. However, the fix in https://github.com/darktable-org/darktable/pull/12233 doesn't cause them in itself, I believe. It just makes the issue more visible due to triggering filmic's reconstruction into generating those blotches.

**Comment by jenshannoschwalm** [comment]
Would there be a way to check in either module for such an issue and give a "module warning"? (BTW I find the black pixels far worse as "not so easy to catch" if not pixelpeeping.)

**Comment by flannelhead** [comment]
> BTW I find the black pixels far worse as "not so easy to catch" if not pixelpeeping.
In the case of this image, yes. But it can happen in vastly larger areas (e.g. sun disc) so that's a no-no anyway (in my opinion).

> Would there be a way to check in either module for such an issue and give a "module warning"?
I'm not sure. The best thing to do would be probably to fix the brilliance formula (or clamping) in such a way that avoids those vast changes of bright highlights and clamps them to some sensible range. That was already briefly discussed in https://github.com/darktable-org/darktable/issues/12163 but would need more discussion because there didn't seem to be a viable fix yet.

**Comment by todd-prior** [comment]
I dont know if it offers any clues but the artifacts do not appear the same at different zooms in the central preview and dont match the preview used for panning and zoom navigation which seems to breifly show during panning on the center display and then disappears and this also is nothing like what gets exported... I checked and got the same with or without Opencl. I posted on the topic with a video and still example..... For sure the strongest impact seems to be the filmic transition slider but the data used being different might indicate more???

**Comment by herrkami** [comment]
I just edited around 200 images and encountered the problem in around 20 of those. In five cases the blobs revealed themself only after export, because of the zoom level dependency.

**Comment by da-phil** [comment]
I can also reproduce this issue on a couple of my images too. The "workaround" described by @flannelhead works though.

**Comment by TurboGit** [comment]
@flannelhead : Do you think you'll have time to handle this for 4.2?

**Comment by flannelhead** [comment]
Not sure yet. If anything, this would deserve a "proper" fix, such that the brilliance formula is modified to give more sensible results. As it currently stands, it's far too easy to push the luminance to a nonsense territory of very very high values.

I can try to take a look at it during November. Failing that, I would suggest just reverting the PR https://github.com/darktable-org/darktable/pull/12233 and accepting the black pixels.

**Comment by da-phil** [comment]
@flannelhead I hope you're not still discouraged from APs words some months ago. Your work is really appreciated and I'm glad there is another sane contributor understanding the colour science magic in dt.

**Comment by flannelhead** [comment]
Not discouraged, been just kind of busy for the whole autumn.

**Comment by Nilvus** [comment]
For future work @flannelhead, #12734 issue is related to this one and can add infos to this one.

**Comment by TurboGit** [comment]
> I can try to take a look at it during November. Failing that, I would suggest just reverting the PR #12233 and accepting the black pixels.

Will do that.

**Comment by gi-man** [comment]
I prefer the white issue that I can notice over the black one that is hard to see. I would vote not to revert.

**Comment by TurboGit** [comment]
@gi-man : That's a sensible view... @flannelhead is you final view a revert? In both cases we have an issue anyway. I have no strong opinion, maybe leaning slightly toward @gi-man option.

**Comment by todd-prior** [comment]
Since Aureliens original defaults were essentially to set this feature at disabled... would there be any harm in setting those to the max ie 6ev threshold  and the transision at the lowest value... as mentioned here 

https://discuss.pixls.us/t/editing-moments-with-darktable/11770/688?u=priort

As it stands to use the feature requires the user to set the slider based on the mask so it would be little extra effort and it might minimize the number of issue with this artifact until it can be fixed....

Just a thought??  I know it more of an avoidance than a fix but still might be useful in the short term..

**Comment by aurelienpierre** [comment]
Don't fix it if you don't understand it. But since idiocracy has become the new way, just break it some more.

It's ok, I gave backups.

**Comment by todd-prior** [comment]
Thanks for the explanation...I think I follow it for the most part. 

Sorry it creates such frustration and anger. I know that you have poured countless hours into the project and are likely the most connected to the color science but I don't think the changes are trying to undo your work just but a cover on the outlet for those of stupid enough to try and stick our fingers in... 

If brilliance can be pushed too far then I think as you say it should be investigated and set in line with the math.

Also the point about addressing the white point as you move along in the masking is a good one and I think you are right its not getting used by a lot of people and this is a good exampe of where that can come into play

I think you have so much to offer but when you are moved to such anger it dilutes your input. I think you can still be strong in your conviction, inform and disagree but what is to be gained by doing it in a barrage of insults.... 

You have forked DT and have your platform to show how it "should" be done so can you not just change that to your liking and communicate your solution ?? You are aware of all the boundries that can break things. Many of us are not so intimately linked to the inner workings of the core modules or able to digest the math as easily as you can so there can be a case for a safety net here and there...

Moving on from that I think the one thing that struck me with this issue is that it was obvious something was likely pushed too hard but the fact that it could also appear or not appear in various zoom levels , lightable previews and or at the time of export was also important.  I believe that the pipelines are managed  differently for center preview and some other things that likely go on for zoomed previews etc but whatever solution was presented it needed to address this aspect as well, ie keep this both managed and consistent when viewing the image. 

I guess an on/off seemed like a staitforward fix ??

**Comment by kofa73** [comment]
@aurelienpierre : thanks for the explanation. Could setting the white threshold be made automatic, so people wouldn't need to pick white manually (and re-pick it after adjusting exposure or using tone EQ)? Or, at least, more visible? When people see the module, they see the easy-to-use-looking sliders first, and have no idea they *have* to set the white level (I actually do that quite often, but it is not something mentioned in the documentation.
See https://docs.darktable.org/usermanual/4.0/en/module-reference/processing-modules/color-balance-rgb/#thresholds:

> Set the white point luminance in EV. This is used to normalize the power setting in the [4 ways](https://docs.darktable.org/usermanual/4.0/en/module-reference/processing-modules/color-balance-rgb/#4-ways-tab) tab.'

I pretty much never use the 4-ways tab; however, based on your explanation, the documentation is wrong.

**Comment by kofa73** [comment]
@aurelienpierre : for the image and XMP posted at https://discuss.pixls.us/t/something-wrong-with-darktable-4-1-0-git272-5a1a1845-1-clipped-highlights/32540, using the white picker sets _white fulcrum_ to +3.23 EV, but the highlights remain blown if highlight's brilliance is set to 22% or above. Setting _white fulcrum_ to +16 EV (the max allowed value) has no effect, either.

**Comment by kofa73** [comment]
Since the white fulcrum must be set correctly for the maths to work, I suggest to:
- remove the picker
- calculate the fulcrum automatically every time the module is processed (when input data changes).

**Comment by Donatzsky** [comment]
Yet another user got in trouble because of this: https://discuss.pixls.us/t/filmic-rgb-with-reconstruction-on-suddenly-burns-through-managed-highlights/38794

It seems to me that the solution would be to fix the UX by soft-limiting the brilliance sliders to ±20%. Then it should even be possible to re-enable filmic HL reconstruction.

Longer term, figuring out a better way to manage the white fulcrum would be worthwhile, but I'm not sure there's an elegant solution to that.

> calculate the fulcrum automatically every time the module is processed (when input data changes).

Wouldn't that lead to unpredictable behaviour? If it doesn't, having an `Automatic white fulcrum` setting, enabling or disabling the current slider, should work.

**Comment by gi-man** [comment]
I think something was done for this issue but not linked here. I don't recall seeing this issue in my images recently. Is it still happening?

**Comment by Donatzsky** [comment]
@gi-man The "fix" was to disable filmic HL reconstruct by default. So as long as you don't use that, you won't notice anything. But as you can see from that pixls thread, it's very much still capable of tripping up unaware users (which would be the majority).

**Comment by jenshannoschwalm** [comment]
This issue is certainly about modules interfering. The best option probably would be to soft-limit brilliance settings but that might still lead to instability of filmic highlights due to other modules "running wild on user request".

**Comment by Donatzsky** [comment]
I have been thinking more about it, and also learned some more about how this works, and I don't think think there's any particularly neat solution.

The biggest obstacle is that the effect of the sliders is cumulative, so changing one slider may limit how far you can go with another. As such, I see three possible solutions:

1. Soft-limit all sliders to 20%. Will limit the risk somewhat, but by no means eliminate it and the user will still not know why things break.
2. Dynamically limit the sliders based on the values of the other sliders. Assuming the math behind this is sound, it should work well. Has the benefit of indicating to the user, in a non-threatening way, that going too far is problematic.
3. Leave the sliders as-is, but show some kind of warning if the cumulative value goes over a threshold. Warnings can be scary to the user, though, so care must be taken in the design.

**Comment by jenshannoschwalm** [comment]
Fixed in master.
